### PR TITLE
testsuite: Introduce Cabal-tests library for common testsuite functions

### LINKED ISSUE #9453: Distribution.Utils.TempTestDir is defined in Cabal but only used in testsuite
I noticed that `Distribution.Utils.TempTestDir` was only used in the
testsuite but defined in the Cabal library. Rather than expose this in
the public interface of the `Cabal` library, it is cleaner to refactor
it into a separate library (`Cabal-tests`) which can be used by any
testsuite component.

Also, in future it gives a clearer place to put utility functions which
need to be shared across the testsuite but not exposed in Cabal.
Cabal-tests can also freely add dependencies (such as exceptions) which
we might want to avoid adding to the Cabal library.

I noticed this when attempting to refactor `withShorterPathForNewBuildStore` function.

### PR DESCRIPTION
**Pull request by mpickering**
I noticed that Distribution.Utils.TempTestDir was only used in the testsuite but defined in the Cabal library. Rather than expose this in the public interface of the `Cabal` library, it is cleaner to refactor it into a separate library (`Cabal-tests`) which can be used by any testsuite component.

Also, in future it gives a clearer place to put utility functions which need to be shared across the testsuite but not exposed in Cabal. Cabal-tests can also freely add dependencies (such as exceptions) which we might want to avoid adding to the Cabal library.

Fixes #9453

### CODE REVIEW

File: changelog.d/issue-9453
Comment (ffaf1) [inline changelog.d/issue-9453:1]: Does this PR impact `cabal` behaviour or `Cabal` users? If that is not the case, a changeloc.d entry is not needed.

Comment (mpickering) [reply changelog.d/issue-9453:1]: I think it affects `Cabal` users as it removes a module from the `Cabal` library. I can remove the entry if you think I should.

Comment (ffaf1) [reply changelog.d/issue-9453:1]: I missed that, thanks. Changelog stays!

**Review by ulysses4ever — APPROVED** [review]
I moved this module to Cabal initially, after discussing the options with @gbaz . We thought that a separate library is a little bit of an overkill, but I agree that it's a cleaner solution. And certainly more sustainable going forward. 

Thanks for putting the effort!

**Comment by michaelpj** [comment]
This could also be an opportunity to dogfood multiple public libraries in Cabal itself by having a `Cabal-testlib` public library in `Cabal` with test utilities?

**Comment by mpickering** [comment]
> This could also be an opportunity to dogfood multiple public libraries in Cabal itself by having a `Cabal-testlib` public library in `Cabal` with test utilities?

I don't see why users of `Cabal` library should use these functions, they seem at home to me in `Cabal-tests` package away from distribution.

**Comment by mpickering** [comment]
Can we have another review please so I can merge this?

**Comment by ulysses4ever** [comment]
I put a needs-review label in the hope that it'll help to find other reviewers. Maybe we should have done it earlier.

**Review by geekosaur — APPROVED** [review]

**Comment by Mikolaj** [comment]
@mpickering: and you can set the merge label ahead of time as well. It also needs a rebase, because GHA settings have changed and the CI in the PR is not in sync with that.

**Comment by mpickering** [comment]
I rebased, and now I have to wait (another) two days before this gets merged?

**Comment by Mikolaj** [comment]
I hope a rebase does not reset the counter, but a comment does. :)

**Comment by ulysses4ever** [comment]
> I hope a rebase does not reset the counter, but a comment does

I think pretty much everything does (Mergify just uses the GitHub API endpoint called `updated`). So, this is a good case for manual intervention, perhaps.

**Comment by Mikolaj** [comment]
@mpickering: shall I manually intervene?

**Pull request merged**
